rep-xxxx.rst

```diff
@@ -0,0 +1,248 @@
+It is the user's responsibility then to make sure that poses are transformed into the appropriate coordinate system before passing them on to such libraries.
+This could be done automatically by the motion planner, or manually before submitting goal poses to the planner.
+
+
+Dual or Multi-arm robots
+------------------------
+
+This REP does not specify any special conventions for robots with multiple arms/groups or kinematic chains.
+The standard approach of prefixing joint and link names to ensure uniqueness of all frames in the (combined) frame hierarchy should be used to avoid collisions.
+See the `Example Frame Hierarchies`_ section for examples of this.
+
+
+Frame Authorities
+-----------------
+
+The frames described in this REP will typically be part of the static description of robot models encoded in urdfs or xacros.
+As such, the frame authority is expected to be an instance of ``robot_state_publisher``, but this is not required.
+In cases where (complicated) kinematics preclude the use of standard nodes, a specialised node capable of publishing the necessary frames could be used.
+
+
+Exceptions
+----------
+
+The scope of potential robotics software is too broad to require all ROS software to follow the guidelines of this REP.
+However, choosing different conventions should be well justified and well documented, and preferably avoided.
+
+
+Example Frame Hierarchies
+=========================
+
+Single manipulator
+------------------
+
+The following shows an example frame hierarchy for a single serial manipulator with just the default tool frame::
+
+  base_link
+  ├ base
+  └ link_1
+    └ link_2
+      └..
+       └ link_N
+         ├ tool0
+         └ flange
+
+Dual manipulator
+----------------
+
+The following shows an example frame hierarchy for a work cell that consists of two manipulators::
+
+  base_link
+  ├ ..
+  ├ left_base_link
+  │ ├ left_base
+  │ └ left_link_1
+  │   └ left_link_2
+  │     └..
+  │      └ left_link_N
+  │        ├ left_tool0
+  │        └ left_flange
+  ├ ..
+  └ right_base_link
+    ├ right_base
+    └ right_link_1
+      └ right_link_2
+        └..
+         └ right_link_N
+           ├ right_tool0
+           └ right_flange
+
+
+Compliance
+==========
+
+This REP depends on and is compliant with REP 103 [#REP103]_, except where stated otherwise.
+
+
+Questions
+=========
+
+#. are there robots that do not have ``tool0`` defined as a child of ``flange`` (I seem to remember some brand having ``tool0`` defined relative to the last link/joint, which was not the flange)? If not: make ``tool0`` child of ``flange``. Otherwise keep ``tool0`` a sibling of ``flange``.
+#. should we discourage naming additional frames ``toolN`` (ie: ``tool1``, ``tool2``, etc)? Such names carry no semantics (frame purpose only known out-of-band), and all names in ROS should be as descriptive as possible.
```

Comment:
@shaun-edwards comment (on ec18f94e2a008a44840d680a8693105b92a3500a):
> Some robots store tool poses in an array.  Tool0 actually refers to the too at index zero, typically defined as all zeros.  Tool1 implies the tool at index 1.

Reply:
There are indeed many mfgs that handle their (tool)frames like that. Are you suggesting we don't discourage `toolN` naming?

I feel our naming should not follow / be (too much) influenced by what industrial controllers do. `tool0` is already unfortunate, as I've seen many people starting to introduce `tool1`, `tool2`, etc. Naming in ROS should be done with semantics in mind, and those names do not have any semantics.

Of course consumers of our models are essentially free to do whatever they want in their own derived models, but I'd really like to recommend against using `toolN` frame names (it's almost like naming a variable `a` and `b`).

Reply:
Note btw that the question is actually wrong (copied old version of comment): in the new setup (with `flange` and `tool0` as separate frames), `tool0` does not get any children. Only `flange` has child frames. So any additional tool frames should be made *siblings* of `tool0` (and they don't even have to have `flange` or 'the last link' as their parent necessarily).

Reply:
In general, I think `tool1`...`toolN` are application tool-frames, and shouldn't be included in the base robot URDF.  But we can note that these names should be reserved for that purpose (representing equivalent robot-controller frames).

My view is that `tool1`...`toolN` are children of `tool0`.  `tool0` typically represents the "all-zeros" tool frame.  By definition, then, any user-defined tool frames are defined relative to the nominal ("all-zeros"/`tool0`) frame.

In reality, these frames are typically used to represent different end-effectors that are attached to the robot flange.  They may be more appropriately represented in ROS as eef-specific TCP frames, not as `tool1`, etc.

Reply:
@JeremyZoss wrote:
> In general, I think `tool1`...`toolN` are application tool-frames, and shouldn't be included in the base robot URDF.  But we can note that these names should be reserved for that purpose (representing equivalent robot-controller frames).
> 
> [..]
> 
> In reality, these frames are typically used to represent different end-effectors that are attached to the robot flange.  They may be more appropriately represented in ROS as eef-specific TCP frames, not as `tool1`, etc.

I agree with your last point: I don't think using (what are essentially) indices for names is a good idea. ROS allows you to use names for these kind of things, and we should use/encourage that.

The question was thus: do people feel strongly about allowing `toolN` names or not? Personally I would like to strongly discourage it, but as I also noted, consumers of our urdfs/xacros can do whatever they want of course.

> My view is that `tool1`...`toolN` are children of `tool0`.  `tool0` typically represents the "all-zeros" tool frame.  By definition, then, any user-defined tool frames are defined relative to the nominal ("all-zeros"/`tool0`) frame.

I'm not sure I share this view. `tool0` just happens to be a tool frame with all entries "at 0". But I could just as easily define another frame that is configured as such, and that wouldn't make it the parent of all other tool frames.

Tool frames are typically stated in manuals as being relative to the "flange". An all-zeros frame happens to lie on the flange for most robots (but this is not guaranteed, IIRC, see the first question).

I believe @shaun-edwards chose the `tool0` name as it has special meaning on ABB controllers (ie: the default one). Fanucs fi don't have that.

Finally: making everything 'after' `tool0` a child of it has its problems (due to it not being REP-103 compliant), which is why `flange` was introduced.